engine/classes/Elgg/Http/ResponseFactory.php

```diff
@@ -0,0 +1,343 @@
+		}
+
+		return $this->prepareResponse($content, $status_code, $headers)->send();
+	}
+
+	/**
+	 * Wraps content for compability with legacy Elgg ajax calls
+	 *
+	 * @param string $content     Response content
+	 * @param string $forward_url Forward URL
+	 * @return string
+	 */
+	public function wrapLegacyAjaxResponse($content = '', $forward_url = REFERRER) {
+
+		$content = $this->stringify($content);
+
+		if ($forward_url === REFERRER) {
+			$forward_url = $this->request->headers->get('Referer');
+		}
+
+		// always pass the full structure to avoid boilerplate JS code.
+		$params = [
+			'output' => '',
+			'status' => 0,
+			'system_messages' => [
+				'error' => [],
+				'success' => []
+			],
+			'current_url' => current_page_url(),
+			'forward_url' => $forward_url,
+		];
+
+		$params['output'] = $this->ajax->decodeJson($content);
+
+		//Grab any system messages so we can inject them via ajax too
+		$system_messages = _elgg_services()->systemMessages->dumpRegister();
+
+		if (isset($system_messages['success'])) {
+			$params['system_messages']['success'] = $system_messages['success'];
+		}
+
+		if (isset($system_messages['error'])) {
+			$params['system_messages']['error'] = $system_messages['error'];
+			$params['status'] = -1;
+		}
+
+		$response_type = $this->parseContext();
+		list($service, $name) = explode(':', $response_type);
+		$context = [
+			$service => $name,
+		];
+		$params = $this->hooks->trigger('output', 'ajax', $context, $params);
+
+		return $this->stringify($params);
+	}
+
+	/**
+	 * Prepares a redirect response
+	 *
+	 * @param string $url         Redirection URL
+	 * @param mixed  $status_code HTTP status code or forward reason
+	 * @return RedirectResponse|void
+	 * @throws SecurityException
+	 */
+	public function redirect($url = REFERRER, $status_code = ELGG_HTTP_FOUND) {
+
+		if ($url === REFERRER) {
+			$url = $this->request->headers->get('Referer');
+		}
+
+		$url = elgg_normalize_url($url);
+
+		// allow plugins to rewrite redirection URL
+		$current_page = current_page_url();
+		$params = [
+			'current_url' => $current_page,
+			'forward_url' => $url
+		];
+		$url = $this->hooks->trigger('forward', (string) $status_code, $params, $url);
+
+		if (!is_int($status_code) || $status_code < 300 || $status_code > 399) {
+			$status_code = ELGG_HTTP_FOUND;
+		}
+
+		return $this->prepareRedirectResponse($url, $status_code)->send();
+	}
+
+	/**
+	 * Parses response type to be used as plugin hook type
+	 * @return string
+	 */
+	public function parseContext() {
+
+		$segments = $this->request->getUrlSegments();
+
+		$identifier = array_shift($segments);
+		switch ($identifier) {
+			case 'ajax' :
+				$page = array_shift($segments);
+				if ($page === 'view') {
+					$view = implode('/', $segments);
+					return "view:$view";
+				} else if ($page === 'form') {
+					$form = implode('/', $segments);
+					return "form:$form";
+				}
+				break;
+
+			case 'action' :
+				$action = implode('/', $segments);
```

Comment:
Let's leave current behavior and think of a way to make actions reusable in another ticket. That would be handy for bulk operations but is irrelevant for this PR